File: cmd/server/routes/graph.go
package routes

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os/exec"
	"strings"
	"sync"
	"time"

	"kubecmds-viz/server/graph"
	"kubecmds-viz/server/store"
)

func registerGraph(mux *http.ServeMux) {
	mux.HandleFunc("GET /api/graph", handleGraph)
}

// GET /api/graph
// Snapshot mode: reads from k8s-snapshot/ YAML files.
// Realtime mode: runs 9 parallel kubectl get -A -o json batches.
func handleGraph(w http.ResponseWriter, r *http.Request) {
	if r.URL.Query().Get("snapshot") == "true" {
		handleGraphSnapshot(w)
		return
	}
	handleGraphRealtime(w, r)
}

// --- Snapshot ---

func handleGraphSnapshot(w http.ResponseWriter) {
	namespaces := store.ListBackupNamespaces()
	getItems := func(ns, resourceKey string) []graph.K8sItem {
		list := store.LoadYaml(resourceKey+".yaml", ns)
		if list == nil {
			return nil
		}
		return list.Items
	}
	result := graph.BuildGraph(getItems, namespaces)
	writeJSON(w, http.StatusOK, result)
}

// --- Realtime ---

// groupOf extracts the API group from an apiVersion ("apps/v1" -> "apps", "v1" -> "").
func groupOf(apiVersion string) string {
	if i := strings.Index(apiVersion, "/"); i >= 0 {
		return apiVersion[:i]
	}
	return ""
}

type batchResult struct {
	items []graph.K8sItem
	err   string
}

func handleGraphRealtime(w http.ResponseWriter, _ *http.Request) {
	res := store.GraphResources()

	// One batch for all built-in types; individual calls for CRDs so a missing
	// CRD doesn't fail the core fetch. Core batch is index 0.
	var builtinTypes, batches []string
	for _, rc := range res {
		if rc.ResourceType != rc.Key { // CRD
			batches = append(batches, rc.ResourceType)
		} else {
			builtinTypes = append(builtinTypes, rc.ResourceType)
		}
	}
	batches = append([]string{strings.Join(builtinTypes, ",")}, batches...)

	// "group/kind" -> resourceKey, so kinds sharing a Kind name across API
	// groups (e.g. Gateway) don't collide.
	groupKindToKey := map[string]string{}
	for _, rc := range res {
		groupKindToKey[rc.Group+"/"+rc.Kind] = rc.Key
	}

	results := make([]batchResult, len(batches))
	var wg sync.WaitGroup
	for i, b := range batches {
		wg.Add(1)
		go func(i int, resources string) {
			defer wg.Done()
			results[i] = fetchBatch(resources)
		}(i, b)
	}
	wg.Wait()

	// Core batch (index 0) failing means kubectl itself is broken.
	if results[0].err != "" {
		writeJSON(w, http.StatusInternalServerError, map[string]any{"message": results[0].err})
		return
	}

	// Build ns → resourceKey → []K8sItem index.
	nsData := map[string]map[string][]graph.K8sItem{}
	allNsSet := map[string]bool{}
	for _, br := range results {
		for _, item := range br.items {
			meta, _ := item["metadata"].(map[string]interface{})
			ns, _ := meta["namespace"].(string)
			if ns == "" {
				ns = "_cluster"
			}
			kind, _ := item["kind"].(string)
			apiVersion, _ := item["apiVersion"].(string)
			key, ok := groupKindToKey[groupOf(apiVersion)+"/"+kind]
			if !ok {
				continue
			}
			allNsSet[ns] = true
			if nsData[ns] == nil {
				nsData[ns] = map[string][]graph.K8sItem{}
			}
			nsData[ns][key] = append(nsData[ns][key], item)
		}
	}

	namespaces := make([]string, 0, len(allNsSet))
	for ns := range allNsSet {
		namespaces = append(namespaces, ns)
	}

	getItems := func(ns, resourceKey string) []graph.K8sItem {
		if m, ok := nsData[ns]; ok {
			return m[resourceKey]
		}
		return nil
	}

	result := graph.BuildGraph(getItems, namespaces)
	writeJSON(w, http.StatusOK, result)
}

// fetchBatch runs a single kubectl get batch and returns parsed items.
func fetchBatch(resources string) batchResult {
	args := []string{"get", resources, "-A", "-o", "json"}
	cmd := exec.Command("kubectl", args...)
	cmd.WaitDelay = 30 * time.Second

	out, err := cmd.Output()
	if err != nil {
		msg := err.Error()
		if ee, ok := err.(*exec.ExitError); ok {
			msg = string(ee.Stderr)
		}
		log.Printf("[graph] kubectl get %s: %s", resources, msg)
		return batchResult{err: msg}
	}

	var list struct {
		Items []graph.K8sItem `json:"items"`
	}
	if err := json.Unmarshal(out, &list); err != nil {
		return batchResult{err: fmt.Sprintf("parse error: %v", err)}
	}

	log.Printf("[graph] kubectl get %s: %d items (%.1fKB)",
		resources, len(list.Items), float64(len(out))/1024)

	return batchResult{items: list.Items}
}

File: api/routes/graph.ts
import express from 'express';
import type { Request, Response } from 'express';
import { execFile } from 'child_process';
import { promisify } from 'util';

import { discoverNamespaces, getItemsFromSnapshot, buildGraph } from '../utils/graph-builder';
import type { K8sItem } from '../utils/snapshot-loader';
import { getGraphResources, isCrd } from '../utils/config-loader';
import { snapshotDir } from '../utils/paths';

const execFileAsync = promisify(execFile);

const router = express.Router();

/** A failed execFile carries the child's output and an abort code on the error. */
interface ExecError extends Error {
  code?: string;
  stderr?: string;
}

/** Only the shape this route reads out of `kubectl get -o json`. */
interface KubectlList {
  items?: K8sItem[];
}

// --- Realtime (kubectl) helpers ---

async function execKubectl(args: string, signal: AbortSignal): Promise<{ data: KubectlList; error: string | null }> {
  try {
    const argList = args.split(/\s+/);
    const { stdout } = await execFileAsync('kubectl', argList, {
      encoding: 'utf8',
      timeout: 30000,
      maxBuffer: 50 * 1024 * 1024,
      signal,
    });
    const bytes = Buffer.byteLength(stdout, 'utf8');
    const parsed: KubectlList = JSON.parse(stdout);
    console.log(`[graph] kubectl ${args.split(' -')[0]}: ${(bytes / 1024).toFixed(1)}KB, ${parsed.items?.length ?? 0} items`);
    return { data: parsed, error: null };
  } catch (err) {
    const e = err as ExecError;
    if (e.code === 'ABORT_ERR') return { data: { items: [] }, error: null };
    const stderrLines = [...new Set((e.stderr || '').split('\n').map(l => l.trim()).filter(Boolean))];
    const msg = stderrLines.join('\n') || e.message?.split('\n')[0] || 'Unknown error';
    console.warn(`[graph] kubectl ${args}: ${msg}`);
    return { data: { items: [] }, error: msg };
  }
}

async function fetchLiveData(signal: AbortSignal) {
  const graphResources = getGraphResources();
  const builtins = graphResources.filter(r => !isCrd(r));
  const crds = graphResources.filter(r => isCrd(r));

  // One batch for all built-in types; individual calls for CRDs so a missing
  // CRD doesn't fail the core fetch.
  const batches = [
    builtins.map(r => r.resourceType).join(','),
    ...crds.map(r => r.resourceType),
  ];

  // "group/kind" → internal key, so kinds sharing a Kind name across API groups
  // (e.g. Gateway in gateway.networking.k8s.io vs networking.istio.io) don't collide.
  const groupKindToKey: Record<string, string> = {};
  for (const r of graphResources) groupKindToKey[`${r.group}/${r.kind}`] = r.key;

  const nsData = new Map<string, Map<string, K8sItem[]>>();
  const allNamespaces = new Set<string>();

  function ingest(data: KubectlList) {
    for (const item of data?.items || []) {
      const av = item.apiVersion || '';
      const group = av.includes('/') ? av.slice(0, av.indexOf('/')) : '';
      const key = groupKindToKey[`${group}/${item.kind}`];
      if (!key) continue;
      const ns = item.metadata?.namespace || '_cluster';
      allNamespaces.add(ns);
      if (!nsData.has(ns)) nsData.set(ns, new Map());
      const nsMap = nsData.get(ns)!;
      if (!nsMap.has(key)) nsMap.set(key, []);
      nsMap.get(key)!.push(item);
    }
  }

  const results = await Promise.all(
    batches.map(resources => execKubectl(`get ${resources} -A -o json`, signal))
  );

  for (const r of results) ingest(r.data);

  // Core batch (index 0) failing means kubectl itself is broken.
  if (results[0]?.error) throw new Error(results[0].error);

  return { nsData, namespaces: [...allNamespaces] };
}

// GET /api/graph
router.get('/graph', async (req: Request, res: Response) => {
  const isSnapshot = req.query.snapshot === 'true';

  try {
    if (isSnapshot) {
      // One helper for the whole app, so the export, the loader and this route
      // cannot disagree about where a snapshot lives. It already honours
      // K8S_SNAPSHOT_PATH.
      const dataPath = snapshotDir();

      const namespaceDirs = discoverNamespaces(dataPath);
      const namespaceList = [...namespaceDirs.keys()];

      const getItemsFn = (ns: string, resourceKey: string) => {
        const nsDir = namespaceDirs.get(ns);
        if (!nsDir) return [];
        return getItemsFromSnapshot(nsDir, resourceKey);
      };

      res.json(buildGraph(getItemsFn, namespaceList));
    } else {
      // Abort kubectl processes if client disconnects (e.g. mode switch)
      const ac = new AbortController();
      req.on('close', () => {
        if (!res.writableFinished) {
          console.log('\x1b[31m[graph] Client disconnected — aborting kubectl processes\x1b[0m');
          ac.abort();
        }
      });

      const { nsData, namespaces } = await fetchLiveData(ac.signal);

      const getItemsFn = (ns: string, resourceKey: string) => {
        const nsMap = nsData.get(ns);
        if (!nsMap) return [];
        return nsMap.get(resourceKey) || [];
      };

      res.json(buildGraph(getItemsFn, namespaces));
    }
  } catch (err) {
    const e = err as Error;
    console.error('[graph] Error:', e.message);
    res.status(500).json({ message: e.message || 'Failed to fetch graph data' });
  }
});

export = router;
